src/Stratis.Bitcoin.Features.BlockStore/BlockStore.cs

```diff
@@ -0,0 +1,363 @@
+
+                if (this.storeSettings.TxIndex)
+                    await this.blockRepository.SetTxIndexAsync(this.storeSettings.TxIndex).ConfigureAwait(false);
+            }
+            
+            // Make sure that block store initializes before the consensus. 
+            // This is needed in order to rewind consensus in case it is initialized ahead of the block store.
+            if (this.chainState.ConsensusTip != null)
+            {
+                this.logger.LogCritical("Block store initialized after the consensus!");
+                throw new BlockStoreException("Block store initialized after consensus!");
+            }
+
+            // Start dequeuing.
+            this.currentBatchSizeBytes = 0;
+            this.dequeueLoopTask = this.DequeueBlocksContinuouslyAsync();
+
+            this.logger.LogTrace("(-)");
+        }
+        
+        /// <summary>Sets the store tip.</summary>
+        private void SetStoreTip(ChainedHeader newTip)
+        {
+            this.StoreTip = newTip;
+            this.chainState.BlockStoreTip = newTip;
+        }
+
+        /// <summary>
+        /// Sets block store tip to the last block that exists both in the repository and in the <see cref="ConcurrentChain"/>.
+        /// </summary>
+        private async Task RecoverStoreTipAsync()
+        {
+            this.logger.LogTrace("()");
+
+            var blockStoreResetList = new List<uint256>();
+            Block resetBlock = await this.blockRepository.GetAsync(this.blockRepository.BlockHash).ConfigureAwait(false);
+            uint256 resetBlockHash = resetBlock.GetHash();
+
+            while (this.chain.GetBlock(resetBlockHash) == null)
+            {
+                blockStoreResetList.Add(resetBlockHash);
+
+                if (resetBlock.Header.HashPrevBlock == this.chain.Genesis.HashBlock)
+                {
+                    resetBlockHash = this.chain.Genesis.HashBlock;
+                    break;
+                }
+
+                resetBlock = await this.blockRepository.GetAsync(resetBlock.Header.HashPrevBlock).ConfigureAwait(false);
+                Guard.NotNull(resetBlock, nameof(resetBlock));
+                resetBlockHash = resetBlock.GetHash();
+            }
+
+            ChainedHeader newTip = this.chain.GetBlock(resetBlockHash);
+            await this.blockRepository.DeleteAsync(newTip.HashBlock, blockStoreResetList).ConfigureAwait(false);
+
+            this.SetStoreTip(newTip);
+
+            this.logger.LogWarning("Block store tip recovered to block '{0}'.", newTip);
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>
+        /// Adds a block to the saving queue.
+        /// </summary>
+        /// <param name="blockPair">The block and its chained header pair to be added to pending storage.</param>
+        public void AddToPending(BlockPair blockPair)
```

Comment:
Could this be AddToPendingQueue?

Reply:
Method name was inherited from prev implementation. 
I thing it's better than `AddToPendingQueue` because queue is not a public member. Caller of the `AddToPending` shouldn't even know what data structure we are using internally.

Reply:
Ah ok, that makes sense

Reply:
In that case the caller should not even know the block is pending (so it should be AddBlock or PushBlock) but AddToPendingQueue is not a bad idea it will give the caller an indication that the block is still pending and is pushed to a queue, perhaps even remove Pending if its in a queue its obviously pending.

Reply:
I think if we call the class BockStoreQueue then we can leave the name as `AddToPending`